### PR TITLE
[#488] Sync using id types of neofs sdk

### LINKED ISSUE #488: Passing and returning ID types of NeoFS SDK
Define how we handle basic SDK types in code, by pointer, by value, mixed. Sync it everywhere in the code.

Types:
- `oid.ID`
- `cid.ID`
- `oid.Address`
- `user.ID`

---

It depends on the approach taken: return `(T, error)` or `(*T, error)`. IMO first is expected to improve performance (but it's compilator-dependent) while latter is easier to read. We have same discussion in neofs-node https://github.com/nspcc-dev/neofs-node/pull/1418#discussion_r880087590.

I suggest you also discuss, if you decide to return by value - let me know and I'll change.

_Originally posted by @cthulhu-rider in https://github.com/nspcc-dev/neofs-s3-gw/pull/458#discussion_r882371570_

### PR DESCRIPTION
**Pull request by KirillovDenis**
close #488

### CODE REVIEW

File: api/data/tree.go [hunk 1]
```diff
@@ -75,9 +75,62 @@ func (p *PartInfo) ToHeaderString() string {
...
-	RetentionOID *oid.ID
-	UntilDate    string
-	IsCompliance bool
+	id uint64
+
+	legalHoldOID oid.ID
+	setLegalHold bool
+
+	retentionOID oid.ID
+	setRetention bool
```
Comment (masterSplinter01) [inline api/data/tree.go:84]: I thought you decided to group fields like
```
id

legalHoldOID
setLegalHoldOID

retentionOID
setRetention

untilDate
isCompliance
```

Shouldn't be empty line placed after `setRetention` or line 82 removed?

File: api/layer/cors.go [hunk 2]
```diff
@@ -8,6 +8,8 @@ import (
 	"fmt"
 	"io"
 
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+
```
Comment (masterSplinter01) [inline api/layer/cors.go:12]: empty line

File: api/data/tree.go [hunk 3]
```diff
@@ -75,9 +75,62 @@ func (p *PartInfo) ToHeaderString() string {
...
+	return l.id
+}
+
+func (l *LockInfo) SetLegalHold(objID oid.ID) {
+	l.legalHoldOID = objID
+	l.setLegalHold = true
+}
+
+func (l *LockInfo) UnSetLegalHold() {
+	l.legalHoldOID = oid.ID{}
```
Comment (alexvanin) [inline api/data/tree.go:103]: It frees one object but creates another. Maybe flag update is enough?

File: api/data/tree.go [hunk 4]
```diff
@@ -75,9 +75,62 @@ func (p *PartInfo) ToHeaderString() string {
...
+
+func (l *LockInfo) Retention() oid.ID {
+	return l.retentionOID
+}
+
+func (l *LockInfo) UntilDate() string {
+	return l.untilDate
+}
+
+func (l *LockInfo) IsCompliance() bool {
```
Comment (alexvanin) [inline api/data/tree.go:134]: Getters can be implemented on struct receivers instead of pointer receivers.

**Review by alexvanin — COMMENTED** [review]
By the way, I am okay with keeping pointers on optional IDs. Nil check looks a bit more appealing for me than compassion with empty ID, even though it safer.

File: api/data/tree.go [hunk 5]
```diff
@@ -75,9 +75,62 @@ func (p *PartInfo) ToHeaderString() string {
...
+func (l *LockInfo) ID() uint64 {
+	return l.id
+}
+
+func (l *LockInfo) SetLegalHold(objID oid.ID) {
+	l.legalHoldOID = objID
+	l.setLegalHold = true
+}
+
+func (l *LockInfo) UnSetLegalHold() {
```
Comment (cthulhu-rider) [inline api/data/tree.go:102]:
```suggestion
func (l *LockInfo) ResetLegalHold() {
```

File: api/layer/cors.go [hunk 6]
```diff
@@ -48,13 +50,13 @@ func (n *layer) PutBucketCORS(ctx context.Context, p *PutCORSParams) error {
...
 
-	objIDToDelete, err := n.treeService.PutBucketCORS(ctx, &p.BktInfo.CID, objID)
+	objIDToDelete, err := n.treeService.PutBucketCORS(ctx, p.BktInfo.CID, objID)
 	if err != nil {
 		return err
 	}
 
-	if objIDToDelete != nil {
-		if err = n.objectDelete(ctx, p.BktInfo, *objIDToDelete); err != nil {
+	if !objIDToDelete.Equals(oid.ID{}) {
```
Comment (cthulhu-rider) [inline api/layer/cors.go:58]: I think such an expected behavior should be reflected in `TreeService.PutBucketCORS` docs. I also suggest to not compare ID with zero value, but return some declared error instead (docs too).

Same for `DeleteBucketCORS`, `AddPart` and all other similar places.

File: api/layer/neofs.go [hunk 7]
```diff
@@ -148,7 +148,7 @@ type NeoFS interface {
 	//
 	// It returns exactly one non-nil value. It returns any error encountered which
```
Comment (cthulhu-rider) [inline api/layer/neofs.go:149]:
```suggestion
	// It returns exactly one non-zero value. It returns any error encountered which
```

Same for `CreateObject`.

File: api/layer/system_object.go [hunk 8]
```diff
@@ -36,47 +36,49 @@ func (n *layer) PutLockInfo(ctx context.Context, objVersion *ObjectVersion, newL
...
+			if lockInfo.IsCompliance() {
 				return fmt.Errorf("you cannot change compliance mode")
 			}
 			if !newLock.Retention.ByPassedGovernance {
 				return fmt.Errorf("you cannot bypass governence mode")
 			}
 
-			if len(lockInfo.UntilDate) > 0 {
-				parsedTime, err := time.Parse(time.RFC3339, lockInfo.UntilDate)
+			if len(lockInfo.UntilDate()) > 0 {
```
Comment (cthulhu-rider) [inline api/layer/system_object.go:47]: Lets call `UntilDate()` once?

File: api/layer/tree_service.go [hunk 9]
```diff
@@ -12,58 +12,58 @@ import (
...
-	GetMultipartUploadsByPrefix(ctx context.Context, cnrID *cid.ID, prefix string) ([]*data.MultipartInfo, error)
-	GetMultipartUpload(ctx context.Context, cnrID *cid.ID, objectName, uploadID string) (*data.MultipartInfo, error)
+	CreateMultipartUpload(ctx context.Context, cnrID cid.ID, info *data.MultipartInfo) error
+	DeleteMultipartUpload(ctx context.Context, cnrID cid.ID, multipartNodeID uint64) error
+	GetMultipartUploadsByPrefix(ctx context.Context, cnrID cid.ID, prefix string) ([]*data.MultipartInfo, error)
+	GetMultipartUpload(ctx context.Context, cnrID cid.ID, objectName, uploadID string) (*data.MultipartInfo, error)
 
 	// AddPart puts a node to a system tree as a child of appropriate multipart upload
 	// and returns objectID of a previous part which must be deleted in NeoFS.
 	// If a part is being added for the first time, the previous part ID will be nil.
```
Comment (cthulhu-rider) [inline api/layer/tree_service.go:59]: `the previous part ID will be nil` need to be changed, too. It's similar to already mentioned place of interface usage.

File: authmate/session_tokens_test.go [hunk 10]
```diff
@@ -28,9 +29,9 @@ func TestContainerSessionRules(t *testing.T) {
 
 	require.Len(t, sessionContext, 3)
 	require.Equal(t, sessionContext[0].verb, session.VerbContainerPut)
-	require.Nil(t, sessionContext[0].containerID)
+	require.Equal(t, cid.ID{}, sessionContext[0].containerID)
```
Comment (cthulhu-rider) [inline authmate/session_tokens_test.go:32]:
```suggestion
	require.Zero(t, sessionContext[0].containerID)
```

Same below.

File: creds/tokens/credentials.go [hunk 11]
```diff
@@ -55,7 +55,7 @@ type NeoFS interface {
 	//
 	// It returns exactly one non-nil value. It returns any error encountered which
```
Comment (cthulhu-rider) [inline creds/tokens/credentials.go:56]:
```suggestion
	// It returns exactly one non-zero value. It returns any error encountered which
```

File: api/data/tree.go (hunk 1 above)
Comment (KirillovDenis) [reply api/data/tree.go:84]: Actually not. I wanted to group fields such way to show that `untilDate` and `isCompliance` are tough related to `retentionOID`

**Pull request merged**